### PR TITLE
[Feature] Fix multi-target files

### PR DESCRIPTION
**Pull request by Tritlo**
Doesn't quite work yet, some of the slow tests (the genetic ones with multiple fixes) still fail. But if you run it on a module with fixes in multiple identifiers, it works!

### CODE REVIEW

**Comment by lapplislazuli** [comment]
Current VERBOSE output is 

```
Expr_Ex  | final cmd is /app/endemic /input/Expr.hs --log-level=VERBOSE  --log-file=/output/docker-endemic.log  --config=/config/docker_config.json 
Expr_Ex  | 
Expr_Ex  | <interactive>:1316:5: error: parse error on input ‘=’
Expr_Ex  | endemic: parse failed in: `let
Expr_Ex  |   fake_target_add
Expr_Ex  |     = (let
Expr_Ex  |          add :: Expr -> Expr -> Expr
Expr_Ex  |          cos :: Expr -> Expr
[... Here come alle other type declarations ...]
Expr_Ex  |          x :: Expr
Expr_Ex  |          size exp
Expr_Ex  |            = case exp of
Expr_Ex  |                Num _ -> 0
Expr_Ex  |                Arith _ exp1 exp2 -> 1 + size exp1 + size exp2
Expr_Ex  |                Func _ exp -> 1 + size exp
Expr_Ex  |                Var -> 0
Expr_Ex  |          x = Var
[... Here come all other declarations ... ]
Expr_Ex  |   fake_target_cos
Expr_Ex  |     = (let
Expr_Ex  |          add :: Expr -> Expr -> Expr
Expr_Ex  |          cos :: Expr -> Expr
[ ... Repeat for all other toplevel defs ... ]
Expr_Ex  | in _`
Expr_Ex  | CallStack (from HasCallStack):
Expr_Ex  |   error, called at src/Endemic/Eval.hs:156:36 in Endemic-0.5.0-inplace:Endemic.Eval
Expr_Ex  |   parseExprNoInit, called at src/Endemic/Eval.hs:151:3 in Endemic-0.5.0-inplace:Endemic.Eval
```

**Comment by Tritlo** [comment]
@lapplislazuli Could you paste the full output? Since it's a parse error, I need all of it to make sense of it.

**Comment by lapplislazuli** [comment]
It does a lot more now, I will have a longer look (also because of some of the stuff I did is not working :O, issues with Prelude sin)

**Comment by lapplislazuli** [comment]
Current state has another issue with imports: 

```
PropsExample_1  | endemic: module ‘Parsing’ is defined in multiple files: /input/Parsing.hs
PropsExample_1  |                                                /input/Parsing.hs
```

(File obviously only exists once)

**Comment by Tritlo** [comment]
Working on all of this :)

**Comment by lapplislazuli** [comment]
I am currently debugging a bit, and there are some anomalies I have already found. 

On broken module the log is the following: 
```
Endemic_Example_1  | 2021-07-29 06:47:54 INFO: Starting Genetic Search at 2021-07-29 06:47:54.9144845 UTC
Endemic_Example_1  | 2021-07-29 06:47:54 INFO: Running 50 Generations with a population of 64
Endemic_Example_1  | 2021-07-29 06:47:56 INFO: Finished Generation 0 at 2021-07-29 06:47:56.3506327 UTC(3 Results)
Endemic_Example_1  | 2021-07-29 06:47:56 INFO: Finished Generation 1 at 2021-07-29 06:47:56.3607509 UTC(0 Results)
Endemic_Example_1  | 2021-07-29 06:47:56 INFO: Finished Generation 2 at 2021-07-29 06:47:56.3666286 UTC(0 Results)
Endemic_Example_1  | 2021-07-29 06:47:56 INFO: Finished Generation 3 at 2021-07-29 06:47:56.3716767 UTC(0 Results)
Endemic_Example_1  | 2021-07-29 06:47:56 INFO: Finished Generation 4 at 2021-07-29 06:47:56.37674 UTC(0 Results)
Endemic_Example_1  | 2021-07-29 06:47:56 INFO: Finished Generation 5 at 2021-07-29 06:47:56.3821013 UTC(0 Results)
[...]
Endemic_Example_1  | 2021-07-29 06:47:56 INFO: Finished Generation 49 at 2021-07-29 06:47:56.5586283 UTC(0 Results)
```

So all except the first generation have no results. 
This matches somewhat three fixes - there no result is found in the first generation, and after that it doesn't make progress.

I put some more DebugStrls in and the following methods are called / not called: 

- performMutation ✅ 
- performCrossover ✅ 
- performEnvironmentSelection ✅ 
- Fittest 👎  (Ok, I am stupid this is only for tournament)

Hence in .Search I think everything seems to be properly run. 

In GenMonad: 

- CoinToss ✅ 
- Initial Population ✅  also twice 👍   (I expected 2, after the first results are removed and replaced)
- TossCoin ✅ 
- RandomDouble 👎 
- eFixCrossover 👎 
- FitnessMany ✅  twice per Generation (should be one in selection, and one in finding winners)
- MinimizingFixes 👎 
- MutateMany ✅  called one per Generation 

Fitness Many only calls ONCE for the uncached (in total Once), which means that either here is a bug, or that there is no mutation. The one call is followed by a lot of audit info from the ghc side, which looks proper to me. 

MutateMany runs through all steps, but the "selection" is never touched (neither of the Cases are touched).

I have the feeling that the error was introduced with the move to MutateMany from Mutate.

File: src/Endemic/Search/Genetic/GenMonad.hs [hunk 1]
```diff
@@ -65,8 +66,7 @@ instance Chromosome EFix where
         dropped = zipWith drop' to_drop gens
 
         EProb {..} = progProblem
-        prog_at_ty = progAtTy e_prog e_ty
-        n_progs = map (`replaceExpr` prog_at_ty) to_mutate
+        n_progs = map (applyFixToEProg e_prog) to_mutate
         (gen'' : gens') = splitGenList gen'
         selection ((p, pFixes), generation) =
           case pickElementUniform pFixes generation of
```
Comment (lapplislazuli) [inline src/Endemic/Search/Genetic/GenMonad.hs:72]: This one is not touched, neither the Just nor the Nothing

**Comment by Tritlo** [comment]
@lapplislazuli I think it's all read now, with a few extra flags that slow it down but remove the source of the weird `(file locked)` errors, but which also means we need to do a lot more re-compilation. I thik you've already gone through most of it, but maybe have a proper look and then merge it to master. It should now work for the actual tests!

**Comment by lapplislazuli** [comment]
So the Genetic-Search does not fail on the Experiment! It finishes gracefully without results. 

We sometimes get compiler-warnings printed between the generations, like such: 
```
PropsExample_1  | fake_targets/target-8233387907637908815/FakeTargetCheck7-46.hs:177:22: warning: [-Woverlapping-patterns]
PropsExample_1  |     Pattern match is redundant
PropsExample_1  |     In a case alternative: Just (e, _) -> ...
PropsExample_1  | 
```

I think those should be on DEBUG level, but I have no clue where they are thrown. 
I would also like to reduce the initial print of "Program to Repair:" to Debug, as it makes up half the log. 🪵 

However, the *mutateMany* still doesn't call any of the branches in Selection. 
Mutate Many is called in general, but I think the precomputed fixes introduce bad behaviour. 
I made a small branch with the fixes `noisy-genmonad` and you can see it by just running `docker-compose up --build`.

**Comment by Tritlo** [comment]
Those are warnings coming from the compiler, when we compile the checks, and not any `log` call from us. So we can either hide  them completely by changing the flags or have them there, no relegating them to DEBUG 😅

**Pull request merged**